File: Remember_renewal/app/src/main/java/com/solo_dev/remember_renewal/Write_Activity/Fragment/WriteFragment.java
package com.solo_dev.remember_renewal.Write_Activity.Fragment;

import android.app.AlertDialog;

import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.fragment.app.ListFragment;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.bumptech.glide.Glide;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.solo_dev.remember_renewal.R;
import com.solo_dev.remember_renewal.Write_Activity.Adapter.Data_Adapter;
import com.solo_dev.remember_renewal.Write_Activity.DataItem.Data_Write;
import com.solo_dev.remember_renewal.Write_Activity.DataItem.Reported_Data;
import com.solo_dev.remember_renewal.Write_Activity.WriteActivity;

import java.text.Collator;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;


public class WriteFragment extends Fragment {

    SwipeRefreshLayout mRefresh;
    List<Data_Write> write = new CopyOnWriteArrayList<>();
    Data_Adapter adapter_write;
    ListView list_written;
    FirebaseDatabase firebaseDatabase = FirebaseDatabase.getInstance();
    DatabaseReference databaseReference = firebaseDatabase.getReference();

    public WriteFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View inflate = inflater.inflate(R.layout.fragment_write, container, false);


        mRefresh = inflate.findViewById(R.id.swipe);
        list_written = inflate.findViewById(R.id.write_list);
        adapter_write = new Data_Adapter(getActivity(), write);
        list_written.setAdapter(adapter_write);



        mRefresh.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                reading_data();
                mRefresh.setRefreshing(false);
            }
        });

        initialize();


        list_written.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView parent, View v, int position, long id) {

                Data_Write item = (Data_Write) parent.getItemAtPosition(position) ;

                View innerView = getLayoutInflater().inflate(R.layout.details_dialog, null);

                final AlertDialog mDialog = new AlertDialog.Builder(getActivity())
                        .setView(innerView)
                        .create();

                mDialog.getWindow().setBackgroundDrawable(new ColorDrawable(android.graphics.Color.TRANSPARENT));

                TextView title = innerView.findViewById(R.id.title);
                TextView contents = innerView.findViewById(R.id.contents);
                TextView date = innerView.findViewById(R.id.date);
                ImageView img_one = innerView.findViewById(R.id.img_contents);
                RelativeLayout rela = innerView.findViewById(R.id.rela_img);

                title.setText(item.getTitle());
                contents.setText(item.getDate());
                date.setText(item.getContents());

                if(item.getImg_contents() != null){
                    StorageReference storageReference = FirebaseStorage.getInstance().getReference();
                    Glide.with(getActivity()).load(storageReference.child(item.getImg_contents())).into(img_one);
                }else{
                    rela.setVisibility(View.GONE);
                }


                mDialog.show();

            }
        }) ;

        return inflate;
    }
    public String getTime(){
        long now2 = System.currentTimeMillis();
        Date date = new Date(now2);
        SimpleDateFormat CurTimeFormat = new SimpleDateFormat("aa HH:mm");

        return CurTimeFormat.format(date);
    }
    public void initialize(){
        databaseReference.child("write").addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                Data_Write write_data = dataSnapshot.getValue(Data_Write.class);

                if (write_data.isViewing()) {
                    write.add(new Data_Write(write_data.getTitle(), write_data.getDate(), write_data.getContents(), write_data.getImg_contents(), write_data.getLiked(), write_data.getReported(), write_data.getGetkeys(), write_data.isViewing(), write_data.getUsers(),write_data.getDisplay_name()));
                    adapter_write.notifyDataSetChanged();
                    repoerted_out(write_data);
                }
                adapter_write.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {
            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
            }
        });
    }
    public void reading_data() {
        for(Data_Write idx : write){
            write.remove(idx);
        }
        databaseReference.child("write").addChildEventListener(new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String s) {
                Data_Write write_data = dataSnapshot.getValue(Data_Write.class);
                if (write_data.isViewing()) {
                    write.add(new Data_Write(write_data.getTitle(), write_data.getDate(), write_data.getContents(), write_data.getImg_contents(), write_data.getLiked(), write_data.getReported(), write_data.getGetkeys(), write_data.isViewing(), write_data.getUsers(),write_data.getDisplay_name()));
                    adapter_write.notifyDataSetChanged();
                    repoerted_out(write_data);
                }
                adapter_write.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {
            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {
            }
        });
    }


    public void repoerted_out(final Data_Write data) {
        if (data.getReported() > 10) {
            databaseReference.child("write").child(data.getGetkeys()).child("viewing").setValue(false);
        }

    }
}

File: Remember_renewal/presentation/src/main/java/com/solo_dev/remember_final/ui/view/fragment/WriteFragment.kt
package com.solo_dev.remember_final.ui.view.fragment

import android.app.Activity
import android.app.Activity.RESULT_OK
import android.app.AlertDialog
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.solo_dev.remember_final.ui.view.activity.detail.DetailActivity
import com.solo_dev.remember_final.R
import com.simple.data.data.module.FirebaseDataBaseModule
import com.simple.data.data.module.FirebaseStorageModule
import com.simple.data.data.module.UIModule
import com.simple.data.data.data.DataWrite
import com.solo_dev.remember_final.databinding.FragmentWriteBinding
import com.solo_dev.remember_final.ui.adapter.write.WriteAdapter
import com.solo_dev.remember_final.ui.adapter.write.listener.onClickItemListener
import kotlinx.coroutines.*
import java.text.SimpleDateFormat
import java.util.*


class WriteFragment : Fragment() {

    var adapterWrite: WriteAdapter? = null
    var firebaseDatabase = FirebaseDatabase.getInstance()
    var databaseReference = firebaseDatabase.reference
    private lateinit var binding : FragmentWriteBinding
    val SCROLLING_UP = -1
    var selectImage: Button? = null
    var storagePath : String? = null
    var viewing = true
    private var filePath: Uri? = null
    var innerView: View? = null
    private lateinit var imagePreview : ImageView


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        binding = FragmentWriteBinding.inflate(inflater, container, false)
        
        return binding.root
      }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerViewInit()
        initLayout()

    }

    private fun initLayout(){
        databaseReference.child("notice_server").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(p0: DataSnapshot) {
                binding.tvServerNotice.isSelected = true
                binding.tvServerNotice.text = p0.value.toString()
            }

            override fun onCancelled(p0: DatabaseError) {
            }

        })

        binding.swipeLayout.setOnRefreshListener {
            GlobalScope.launch {
                refreshData(requireActivity(), adapterWrite!!)
                delay(1000)
                withContext(Dispatchers.Main) {
                    binding.swipeLayout.isRefreshing = false
                }
            }
        }
    }

    private fun recyclerViewInit() {
        adapterWrite = WriteAdapter(requireActivity(), object : onClickItemListener {
            override fun onClick(position: Int, data: DataWrite) {
                val intent = Intent(requireContext(), DetailActivity::class.java)
                intent.putExtra("title", data.title)
                intent.putExtra("contents", data.date)
                intent.putExtra("dateName", "${data.contents} , ${data.displayName}님")
                if (data.imgContents != null) {
                    intent.putExtra("img", data.imgContents)
                }
                startActivity(intent)
            }
        })

        GlobalScope.launch {
            withContext(Dispatchers.Main) {
                UIModule.progressDialog(requireActivity(), UIModule.ShowCode.REQUEST_SHOW)
            }
            withContext(Dispatchers.IO) {
                getJobOfDatabase()
            }
        }

        binding.writeList.layoutManager = LinearLayoutManager(requireContext())
        binding.writeList.adapter = adapterWrite

        binding.floatingActionButton.setOnClickListener {
            createDialog()
        }


    }
    private fun databaseWithoutDialog() {
        FirebaseDataBaseModule.loadWriteList(requireActivity(), adapterWrite!!)
    }

    private fun getJobOfDatabase() {
        refreshData(requireActivity(), adapterWrite!!)
    }


    private fun createDialog() {
        innerView = layoutInflater.inflate(R.layout.dialog_write, null)

        val sendBtn = innerView!!.findViewById<Button>(R.id.sendButton)
        val edTitle: EditText = innerView!!.findViewById(R.id.writeTitle)
        val edContents: EditText = innerView!!.findViewById(R.id.writeContent)

        imagePreview = innerView!!.findViewById(R.id.loadImage)
        selectImage = innerView!!.findViewById(R.id.select)
        val mDialog = AlertDialog.Builder(requireActivity())
                .setView(innerView)
                .create()

        mDialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
        selectImage!!.setOnClickListener(View.OnClickListener { v: View? ->
            val intent = Intent()
            intent.type = "image/*"
            intent.action = Intent.ACTION_GET_CONTENT
            startActivityForResult(Intent.createChooser(intent, "이미지를 선택하세요."), FirebaseDataBaseModule.CODE_UPLOAD)
        })

        sendBtn.setOnClickListener { v: View? ->
            if (!edTitle.text.toString().isEmpty() && !edContents.text.toString().isEmpty()) {
                val task = databaseReference.child("write").push()
                val user = FirebaseAuth.getInstance().currentUser
                val data = DataWrite(edTitle.text.toString(), edContents.text.toString(), time, storagePath, 0, 0, viewing, user?.email!!, user.displayName!!)
                task.setValue(data)
                storagePath = null

                GlobalScope.launch {
                    withContext(Dispatchers.Main) {
                        UIModule.progressDialog(requireActivity(), UIModule.ShowCode.REQUEST_SHOW)
                    }
                    withContext(Dispatchers.IO) {
                        getJobOfDatabase()
                    }
                }

                mDialog.cancel()
            }
        }
        mDialog.show()
    }


    private val time: String
        get() {
            val now2 = System.currentTimeMillis()
            val date = Date(now2)
            val curTimeFormat = SimpleDateFormat("aa HH:mm")
            return curTimeFormat.format(date)
        }

    override fun onResume() {
        super.onResume()
    }

     override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
         if (requestCode == FirebaseDataBaseModule.CODE_UPLOAD && resultCode == RESULT_OK) {
             filePath = data!!.data
             storagePath = FirebaseStorageModule.uploadFile(imagePreview, filePath, requireActivity(), requireContext())
         }
     }

    private fun refreshData(activity: Activity, adapter: WriteAdapter) {
        val data = ArrayList<DataWrite>()
        CoroutineScope(Dispatchers.Default).async {
            val getDatabaseBoardList: Job = GlobalScope.async {
                FirebaseDatabase.getInstance().reference.child("write").addChildEventListener(object : ChildEventListener {
                    override fun onChildAdded(result: DataSnapshot, s: String?) {
                        val writeData = result.getValue(DataWrite::class.java)!!

                        Log.d("TAG", "tag_onAdded")
                        data.add(0, DataWrite(writeData.title, writeData.date, writeData.contents, writeData.imgContents,
                                writeData.liked, writeData.reported, writeData.viewing, writeData.users, writeData.displayName))
                    }

                    override fun onChildChanged(result: DataSnapshot, s: String?) {
                        Log.d("TAG", "tag_onChanged")
                    }

                    override fun onChildRemoved(result: DataSnapshot) {
                        Log.d("TAG", "tag_onRemoved")
                    }

                    override fun onChildMoved(result: DataSnapshot, s: String?) {
                    }

                    override fun onCancelled(result: DatabaseError) {
                        Log.d("Error", result.message)
                    }
                })
                for (index in data) {
                    if (!index.imgContents!!.contains("images/")) {
                        index.imgContents = null
                    }
                }

            }
            getDatabaseBoardList.join()

            delay(2000)

            if (getDatabaseBoardList.isCompleted) {
                withContext(Dispatchers.Main) {
                    UIModule.progressDialog(activity, UIModule.ShowCode.REQUEST_DISMISS)
                    adapter.setData(data)
                }
            }

            Log.d("TAG_", "inCage : ${getDatabaseBoardList.isActive}, ${getDatabaseBoardList.isCompleted}, ${getDatabaseBoardList.isCancelled}")

            true
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        adapterWrite!!.clear()
    }
}